x-pack/plugins/file_upload/public/importer/geojson_importer/geojson_importer.ts

```diff
@@ -18,161 +17,268 @@ import {
-        reject(
-          new Error(
-            i18n.translate('xpack.fileUpload.fileParser.noFeaturesDetected', {
-              defaultMessage: 'Error, no features detected',
-            })
-          )
-        );
-        return;
+    for (let i = 0; i < rawFeatures.length; i++) {
+      const rawFeature = rawFeatures[i] as Feature;
+      if (!isLastBatch && i === rawFeatures.length - 1) {
+        // Do not process last feature until next batch is read, features on batch boundary may be incomplete.
+        this._prevBatchLastFeature = rawFeature;
+        continue;
       }
 
-      const features: Feature[] = [];
-      const geometryTypesMap = new Map<string, boolean>();
-      let invalidCount = 0;
-      for (let i = 0; i < rawFeatures.length; i++) {
-        const rawFeature = rawFeatures[i] as Feature;
-        if (!rawFeature.geometry || !rawFeature.geometry.type) {
-          invalidCount++;
-        } else {
-          if (!geometryTypesMap.has(rawFeature.geometry.type)) {
-            geometryTypesMap.set(rawFeature.geometry.type, true);
-          }
-          features.push(geoJsonCleanAndValidate(rawFeature));
+      this._totalFeatures++;
+      if (!rawFeature.geometry || !rawFeature.geometry.type) {
+        this._invalidFeatures.push({
+          item: this._totalFeatures,
+          reason: i18n.translate('xpack.fileUpload.geojsonImporter.noGeometry', {
+            defaultMessage: 'Feature does not contain required field "geometry"',
+          }),
+          doc: rawFeature,
+        });
+      } else {
+        if (!this._geometryTypesMap.has(rawFeature.geometry.type)) {
+          this._geometryTypesMap.set(rawFeature.geometry.type, true);
         }
+        this._features.push(geoJsonCleanAndValidate(rawFeature));
       }
+    }
+  }
 
-      const errors: string[] = [];
-      if (invalidCount > 0) {
-        errors.push(
-          i18n.translate('xpack.fileUpload.fileParser.featuresOmitted', {
-            defaultMessage: '{invalidCount} features without geometry omitted',
-            values: { invalidCount },
-          })
-        );
-      }
-      resolve({
-        errors,
-        geometryTypes: Array.from(geometryTypesMap.keys()),
-        parsedGeojson: {
-          type: 'FeatureCollection',
-          features,
-        },
-      });
+  public read(data: ArrayBuffer): { success: boolean } {
+    throw new Error('read(data: ArrayBuffer) not supported, use readFile instead.');
+  }
+
+  protected _createDocs(text: string): CreateDocsResponse {
+    throw new Error('_createDocs not implemented.');
+  }
+}
+
+export function toEsDocs(
+  features: Feature[],
+  geoFieldType: ES_FIELD_TYPES.GEO_POINT | ES_FIELD_TYPES.GEO_SHAPE
+) {
+  const esDocs = [];
+  for (let i = 0; i < features.length; i++) {
+    const feature = features[i];
+    const geometry = feature.geometry as
+      | Point
+      | MultiPoint
+      | LineString
+      | MultiLineString
+      | Polygon
+      | MultiPolygon;
```

Comment:
Should this include `geometry_collection` as well?

Reply:
No, the [existing logic](https://github.com/elastic/kibana/blob/master/x-pack/plugins/file_upload/public/importer/geojson_importer/geojson_importer.ts#L152) does not support geometry_collection since it verifies `geometry` is provided. GeometryCollection does not provide `geometry`. I would consider this a separate issue